File: src/backend/api_utils.py
import requests
import json
import base64
from icecream import ic

# Get token from Spotify API returns the access token
def get_token() -> str:
    with open("api_creds.json") as f:
        creds = json.load(f)
    client_id = creds["client_id"]
    client_secret = creds["client_secret"]
    auth = f"{client_id}:{client_secret}"
    b64_auth = base64.b64encode(auth.encode()).decode()
    auth_url = "https://accounts.spotify.com/api/token"
    auth_headers = {
        'Authorization': f"Basic {b64_auth}"
    }
    auth_data = {
        'grant_type': 'client_credentials'
    }
    auth_response = requests.post(auth_url, headers=auth_headers, data=auth_data)
    return auth_response.json()["access_token"]

# Get songs from Spotify API
# Will take params later for genre, artist, existing playlist, etc.
# For now, gets list of random songs
def find_songs(token: str):
    url = "https://api.spotify.com/v1/search"
    headers = {
        'Authorization': f"Bearer {token}"
    }
    params = {
        "query": "kanye",
        "limit": 50,
        "type": "track",
    }
    response = requests.get(url, headers=headers, params=params)
    songs = []
    for song in response.json()["tracks"]["items"]:
        songs.append((f"{song['name']} by {song['artists'][0]['name']}", song["duration_ms"], song["uri"]))
    return songs

# Match songs to intervals
# Returns the "URIs" string that the add to playlist API needs
def match_songs(songs: list, intervals: list[int]) -> str:
    uris = []
    for interval in intervals:
        min_song = songs[0]
        for song in songs:
            _, min_duration, _ = min_song
            _, duration, _ = song
            if abs(duration - interval) < abs(min_duration - interval):
                min_song = song
        uris.append(min_song[2])
        print(f"Matched {min_song[0]} of length {round(min_song[1] / 60000, 2)} to interval of {round(interval / 60000, 2)} minutes")
        songs.remove(min_song)
        if songs == []:
            break
    return ",".join(uris)


# Makes array with the intervals in ms time from minutes
def fill_workout(int_length: int, rest_length: int, total_length: int) -> list[int]:
    intervals = []
    rest_period = True
    while (sum(intervals) / 60000) < total_length:
        if rest_period:
            intervals.append(rest_length * 60000)
        elif not rest_period:
            intervals.append(int_length * 60000)
        rest_period = not rest_period
    intervals.append((total_length - (sum(intervals) / 60000)) * 60000) # fill the time with a final interval
    return intervals

def main():
    int_length = int(input("Enter the length of your active interval: ")) # in minutes
    rest_length = int(input("Enter the length of your rest interval: ")) # in minutes
    total_length = int(input("Enter the total length of your workout: ")) # in minutes
    workout_times = fill_workout(int_length, rest_length, total_length)
    token = get_token()
    songs = find_songs(token)
    uris = match_songs(songs, workout_times)
    print(uris)


if __name__ == "__main__":
    main()

File: src/backend/src/main/scala/ApiUtils.scala
package HiitBeats

import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.Base64
import play.api.libs.json._
import scala.annotation.tailrec
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.duration._
import scala.concurrent.{Await, Promise}
import scala.io.Source
import scala.language.postfixOps
import scala.util.{Try, Success, Failure}
import sttp.client4._
import sttp.client4.httpclient.HttpClientSyncBackend
import sttp.model.Uri


// Case class to store what we want from a Spotify Song
case class Song(name: String, 
                artist: String, 
                duration: Int, 
                uri: String)

// Object to hold the Spotify API functions
object ApiUtils extends SpotifyApi {

  override def getClientToken: Try[String] = Try {
    val (clientId, clientSecret) = getCredentials("/Users/seancoughlin/projects/hiitbeats/src/backend/api_creds.txt")
    val base64Auth = Base64.getEncoder.encodeToString(s"$clientId:$clientSecret".getBytes)
    val headers = Map(
      "Authorization" -> s"Basic $base64Auth",
      "Content-Type" -> "application/x-www-form-urlencoded"
    )
    val body = "grant_type=client_credentials"

    val response = makeRequest(uri"https://accounts.spotify.com/api/token", headers, Some(body)).get
    val json = parseJson(response)
    (json \ "access_token").as[String]
  }

  override def findSongs(token: String, query: String): List[Song] = {
    val headers = Map("Authorization" -> s"Bearer $token")
    val params = Map("q" -> query, "limit" -> "50", "type" -> "track")
    val uriWithParams: Uri = uri"https://api.spotify.com/v1/search".params(params)

    val response = makeRequest(uriWithParams, headers).get
    val json = parseJson(response)
    val rawSongs: Option[JsArray] = (json \ "tracks" \ "items").asOpt[JsArray]

    def tuplize(e: JsValue): Song = {
      val name: Option[String] = (e \ "name").asOpt[String]
      val artist: Option[String] = (e \ "artists")(0) \ "name" match {
        case JsDefined(JsString(name)) => Some(name)
        case _ => None
      }
      val duration: Option[Int] = (e \ "duration_ms").asOpt[Int]
      val uri: Option[String] = (e \ "uri").asOpt[String]
      (name, artist, duration, uri) match {
        case (Some(name), Some(artist), Some(duration), Some(uri)) => Song(name, artist, duration, uri)
        case _ => Song("Error", "Error", 0, "Error")
      }
    }

    rawSongs match {
      case Some(songs) => songs.value.toList.map(e => tuplize(e))
      case None => List(Song("Error", "Error", 0, "Error"))
    }
  }

  override def fillWorkout(intLength: Int, restLength: Int, totalLength: Int): List[Int] = {
    @tailrec
    def fillWorkoutHelper(intLength: Int, restLength: Int, totalLength: Int, acc: List[Int]): List[Int] = {
      if (totalLength <= 0) acc
      else if (totalLength - intLength <= 0) acc :+ totalLength
      else fillWorkoutHelper(intLength, restLength, totalLength - intLength - restLength, acc :+ (restLength * 60000) :+ (intLength * 60000))
    }
    fillWorkoutHelper(intLength, restLength, totalLength, List())
  }  
  
  override def matchSongs(workout: List[Int], songs: List[Song]): String = {
    def getClosestSong(interval: Int, songs: List[Song]): Song = {
      songs.minBy(s => Math.abs(s.duration - interval))
    }
    @tailrec
    def matchSongsHelper(workout: List[Int], songs: List[Song], acc: String): String = {
      workout match {
        case Nil => acc
        case interval :: rest => 
          val closestSong = getClosestSong(interval, songs)
          println(s"Matching ${interval.toFloat/60000.00} to ${closestSong.name} by ${closestSong.artist} of length ${closestSong.duration.toFloat / 60000.00}")
          matchSongsHelper(rest, songs.filterNot(_ == closestSong), acc + closestSong.uri + ",")
      }
    }
    matchSongsHelper(workout, songs, "").stripSuffix(",")
  }

  override def makeUserPlaylist(userID: String, accessToken: String): String = {
    // For the name of the playlist, get the current date and time
    val now = LocalDateTime.now()
    val formatter = DateTimeFormatter.ofPattern("MM/dd/yyyy HH:mm")
    val formattedDateTime = now.format(formatter)

    val headers = Map("Authorization" -> s"Bearer $accessToken", "Content-Type" -> "application/json")
    val body = Json.obj(
      "name" -> s"HiitBeats Workout $formattedDateTime",
      "description" -> "Playlist for your HIIT workout",
      "public" -> false
    ).toString()

    val uri = uri"https://api.spotify.com/v1/users/$userID/playlists"
    val response = makeRequest(uri, headers, Some(body)).get
    val json = parseJson(response)
    (json \ "id").asOpt[String].getOrElse("Error")
  }

  override def addSongsToPlaylist(playlistID: String, accessToken: String, uris: String): Unit = {
    val headers = Map(
      "Authorization" -> s"Bearer $accessToken",
      "Content-Type" -> "application/json"
    )
    val uriList = uris.split(",").toList
    val body = Json.obj(
      "uris" -> uriList
    ).toString()
    val uri = uri"https://api.spotify.com/v1/playlists/$playlistID/tracks"

    val response = makeRequest(uri, headers, Some(body)).get
    if (response.nonEmpty) {
      println("Playlist updated successfully!")
    } else {
      println("Failed to add songs to the playlist.")
    }
  }
}

